config/configmapprovider/file_watch.go

```diff
@@ -0,0 +1,81 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configmapprovider
+
+import (
+	"context"
+	"errors"
+	"os"
+	"time"
+)
+
+var errNoOp = errors.New("no change")
+
+// watchFile sets up a watch on a filename to detect changes and calls onChange() with a suitable ChangeEvent.
+// The watch is cancelled when the context is done.
+func watchFile(ctx context.Context, filename string, onChange func(*ChangeEvent)) error {
+	var (
+		lastfi os.FileInfo
+	)
+
+	check := func() error {
+		currfi, err := os.Stat(filename)
+		modified := false
+		if err != nil {
+			if os.IsNotExist(err) && lastfi != nil {
+				return errNoOp
+			}
+			return err
+		}
+		if lastfi != nil && (currfi.Size() != lastfi.Size() || !currfi.ModTime().Equal(lastfi.ModTime())) {
+			modified = true
+		}
+
+		lastfi = currfi
+		if modified {
+			return nil
+		}
+
+		return errNoOp
+	}
+
+	// Perform an initial check.
+	err := check()
+	if err != nil && err != errNoOp {
+		return err
+	}
+
+	// Check the file every second and if it's been updated then initiate a config reload.
+	go func() {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				// If check returns a valid event, exit the loop. A new watch will be placed on the next Retrieve()
+				err := check()
+				if err == nil || err != errNoOp {
+					onChange(&ChangeEvent{
```

Comment:
This will reload the config immediately when a change is detected. That's an undesirable behavior since the file may be in the middle of being written and we may read partially written file. It is better to wait for some small amount of time (e.g. 1 second) after the last change to the file and only after that trigger reloading to increase the chance that the entire content of the file is written.

Reply:
ack.

Reply:
addressed by adding a sleep.

Reply:
I don't see where this is addressed, please point me to the code.

Reply:
apologies. i somehow removed it during cleanup. added it back.